libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Version.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# OPTIONS_GHC -fno-warn-redundant-constraints #-}
+
+module Cardano.Ledger.Binary.Version
+  ( -- * Versioning
+    Version,
+    MinVersion,
+    MaxVersion,
+    natVersion,
+    natVersionProxy,
+    mkVersion,
+    allVersions,
+  )
+where
+
+import Data.Proxy
+import GHC.TypeLits
+import Numeric.Natural
+
+--------------------------------------------------------------------------------
+-- Versioned Decoder
+--------------------------------------------------------------------------------
+
+-- | Protocol version number that is used during encoding and decoding. All supported
+-- versions are in the range from `MinVersion` to `MaxVersion`.
+newtype Version = Version Word
+  deriving (Eq, Ord, Show, Enum)
+
+-- | Minimum supported version
+type MinVersion = 1
+
+-- | Maximum supported version
+type MaxVersion = 8
```

Comment:
Why are we using type versions of Nat.  None of the these types seem to be used as indexes, so what does it buy us?

Reply:
This gives us a range of valid versions at the type level, which we can use later to construct `Version` without any call to `error`. In particular `natVersion` and `natVersionProxy` let us do that.